source/leuven/services/openondemand.rst

```diff
@@ -191,136 +241,297 @@ of using any of the other partitions. There are however some limitations on the
-:ref:`R <create_r_conda_env>` conda environment. The second step consists out of effectively
-creating the kernel in your ``$VSC_HOME/.local`` folder, as Jupyter will look for kernels in this location. The following commands should be excecuted from a shell, 
-and only need to be done once for the set-up of each new kernel. This starts with activating your conda
-environment::
+   If you use JupyterLab as explained above, and you need to reproduce your numerical results 
+   at any time in the future, make sure you stick to choosing the same 'Toolchain version'.
+
+If the module-based environment explained above does not provide all packages that you need, then
+it is recommended to manage your custom-made R or Python environments by creating custom Jupyter kernels.
+
+Custom Conda environments
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you have not installed Conda in your account yet, please refer to the
+:ref:`install Miniconda <install_miniconda_python>` page.
+
+To create a custom kernel, first create a  :ref:`Python <create_python_conda_env>` or
+:ref:`R <create_r_conda_env>` Conda environment. The second step consists of effectively
+creating the kernel in your ``$VSC_HOME/.local/share`` folder (which is a defalt value for the
+``$XDG_DATA_HOME`` environment variable), as Jupyter will look for kernels in this location.
+The following commands should be excecuted from a shell (e.g. using 'Login Server Shell Access'), and only need
+to be done once for the set-up of each new kernel.
+If you already have an existing Python kernel, but your JupyterLab session freezes/craches when choosing your
+old kernel, you also need to repeat the following steps only once.
+This starts with activating your Conda environment::
 
       source activate <env_name>
 
-For Python you will need the ``ipykernel`` package installed in your conda environment. Then you create the kernel as follows::
+For Python you will need the ``ipykernel`` package installed in your Conda environment.
+The minimum supported version for Python is 3.7 and for ``ipykernel`` package is 6.19.2::
+
+      conda install ipykernel
+
+Then you create the kernel as follows::
 
-      python -m ipykernel install --user --name '<env_name>' --display-name '<kernel_name>'
+      python -m ipykernel install --user --env PYTHONPATH "" --name '<env_name>' --display-name '<kernel_name>'
 
-For R, you need both the ``jupyter_client`` and the ``irkernel`` package installed. With the following command you can create the kernel::
+
+Custom virtual environments for Python
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+One may create a customized kernel starting from a Python virtual environment, using one of Python modules.
+Eventhough this approach can have advantages, but we do not recommend it, due to the fact that Python
+virtual environments are architecture specific.
+Hence, on our current Tier-2 machines, this approach needs to be handled with care.
```

Comment:
`due to the fact that Python virtual environments are architecture specific.`  -> this is not the case. Just as with Conda environments, it mostly depends on how you install the packages. If you're installing precompiled packages (e.g. from PyPI repos via `pip install ...` or from Conda repos via `conda install ...`), then these will have been compiled in such a way that compiled parts (if any) can run on any (x86) architecture (except perhaps very old ones).

When using Python installations from a module, then of course these Python modules are replicated for the different available architectures (Skylake, Cascadelake, ...). In practice, however, it's no problem to create a virtual environment with some `Python/3.X.Y-...` module on one architecture (say, Skylake) and then use this venv on another architecture (say, Genoa). At least that's what I've experienced so far. It's of course best to load `Python/3.X.Y-...` module for the architecture at hand (and not that of another architecture), but that's already how the module-based-JupyterLab-app is operating.

Also: while I wouldn't say we should explicitly recommend virtual environments (nor Conda environments, for that matter), I don't think we should discourage it either.

Reply:
I do not think so. Let's work it out by a way of example.
I have created a kernel from a venv, starting from `Python/3.11.3-GCCcore-12.3.0` on the wICE `interactive` partition. As a result of that, the first lines of my `kernel.json` read:
```
{
 "argv": [
  "/vsc-hard-mounts/leuven-data/307/vsc30745/tests/ondemand/jupyterlab/test_venv_kernel/bin/python",
...
```
when I look up the `python` exactuable pointed at above, I see that it is a softlink pointing at the Python executable on IceLake nodes:
```
ls -l /vsc-hard-mounts/leuven-data/307/vsc30745/tests/ondemand/jupyterlab/test_venv_kernel/bin/python
lrwxrwxrwx 1 vsc30745 vsc30745 82 Jun 14 10:40 /vsc-hard-mounts/leuven-data/307/vsc30745/tests/ondemand/jupyterlab/test_venv_kernel/bin/python -> /apps/leuven/rocky8/icelake/2023a/software/Python/3.11.3-GCCcore-12.3.0/bin/python
```
Now, if I start a job on any architecture other than IceLake nodes, my kernel wil l freeze/crach (I tested this and I saw it).

The reason for the issue is not the extra packages that we install with `pip` into our venv, but is the non-portability of the architecture-specific binaries.

Reply:
Perhaps this can be explained more correctly then. I would omit the part that says that it is architecture-specific, but rather say something like:

'If you want to use Python venvs, be aware that you can only use that venv for the specific architecture you created it on, as the kernel will link to the specific Python module/version it was created with. This means that on other architectures, that module won't be available, which will result in the crashing of your kernel.'

Reply:
Ehsan: could you describe your example in more detail? Because I myself cannot reproduce it (i.e. so far I haven't had problems of that sort), perhaps my own test is too simple:

Just now, I created a venv based on a Python/3.9.5-... module from `icelake/2021a` on a wICE interactive node in which I installed NumPy and then created the kernel for it. I then used OOD in test to launch JupyterLab (with 2021a toolchain version; no scriptlet) on the test batch partition, so with skylake-6132 CPUs. I could just select my kernel for a new notebook and e.g. calculate some dot products with NumPy. And just to emphasize: in this situation, the Python interpreter used is the skylake-6132-installed one (which is indeed a requirement for this to work).

Reply:
I admit that it is possible to create venv on one arch and use it on another arch, as I verified on both sandbox skylake nodes and wice sapphire rapids nodes.
The commit 9808326 shall reflect the changes to the text under the venv section.